File: src/admin/port.ts
import type {
  AdminCapabilities,
  Customer,
  CustomerInput,
  CustomerListQuery,
  Feature,
  FeatureInput,
  Meter,
  MeterInput,
  Plan,
  PlanInput,
  Subscription,
  SubscriptionInput,
} from "./types.js";

export interface OpenMeterAdmin {
  readonly capabilities: AdminCapabilities;
  waitForHealthy(): Promise<void>;
  listMeters(): Promise<Meter[]>;
  createMeter(input: MeterInput): Promise<Meter>;
  listFeatures(): Promise<Feature[]>;
  createFeature(input: FeatureInput): Promise<Feature>;
  listPlans(): Promise<Plan[]>;
  ensurePlan(input: PlanInput): Promise<Plan>;
  listCustomers(query?: CustomerListQuery): Promise<Customer[]>;
  createCustomer(input: CustomerInput): Promise<Customer>;
  listCustomerSubscriptions(customerId: string): Promise<Subscription[]>;
  createSubscription(input: SubscriptionInput): Promise<Subscription>;
}

File: internal/admin/port.go
package admin

import "context"

type OpenMeterAdmin interface {
	WaitForHealthy(ctx context.Context) error
	ListMeters(ctx context.Context) ([]Meter, error)
	CreateMeter(ctx context.Context, input MeterInput) (*Meter, error)
	DeleteMeter(ctx context.Context, id string) error
	ListFeatures(ctx context.Context) ([]Feature, error)
	CreateFeature(ctx context.Context, input FeatureInput) (*Feature, error)
	DeleteFeature(ctx context.Context, id string) error
	ListPlans(ctx context.Context) ([]Plan, error)
	EnsurePlan(ctx context.Context, input PlanInput) (*Plan, error)
	DeletePlan(ctx context.Context, id string) error
}
